File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/HomeShelvesRow.swift
internal import SwiftUI

internal import DesignSystem
internal import Persistence
internal import StoryArcCore

/// One of the home surface's two shelves of shelves.
///
/// `collections-and-reading-lists`, *Shelves on the home surface*. The caller draws it only
/// when its half holds something — a heading over a gap is the surface looking like it is
/// waiting for something, which is the one thing it must never do.
///
/// **The heading leads to ``ShelvesView``, not to a filtered library.** Every other heading
/// here leads to "the full list in the library, filtered to match the shelf", and a collection
/// is not a library filter: the exhaustive list of collections is a screen of its own, and it
/// is the only place a shelf is made, renamed or deleted. `HomeScreen`'s `pinnedSections`
/// makes the same argument from the other side, where the shelf is a collection's *contents*.
///
/// Android's `homeShelvesShelf` is the same shelf.
struct HomeShelvesRow: View {
    let title: LocalizedStringKey
    let summaries: [HomeShelfSummary]
    let model: LibraryModel
    /// The Kavita servers this device can open, by source id — see ``KavitaPage/pages(in:)``.
    let pages: [UUID: KavitaPage]
    let onOpen: (Publication, URL) -> Void

    var body: some View {
        HomeSection(title: Text(title, bundle: .module)) {
            ShelvesView(model: model, onOpen: onOpen)
        } content: {
            row
        }
    }

    private var row: some View {
        ScrollView(.horizontal) {
            LazyHStack(alignment: .top, spacing: StoryArcSpace.rowCoverGap) {
                ForEach(summaries) { summary in
                    NavigationLink {
                        destination(of: summary)
                    } label: {
                        card(summary)
                    }
                    .buttonStyle(.plain)
                    // A shelf is a composite of four covers, and four covers below about 150
                    // points stop being four covers — the floor `ShelvesView`'s own grid
                    // takes, rather than a cover cell's.
                    .frame(width: 150)
                }
            }
            .scrollTargetLayout()
        }
        .scrollIndicators(.hidden)
        .scrollTargetBehavior(.viewAligned)
        .contentMargins(.horizontal, StoryArcSpace.gutter, for: .scrollContent)
    }

    /// The card itself: ``ShelfCard``, unchanged, because this surface and the shelves screen
    /// draw the same thing and a second card would eventually be a different one.
    private func card(_ summary: HomeShelfSummary) -> some View {
        ShelfCard(
            model: model,
            title: summary.name,
            subtitle: shelfSubtitle(count: summary.count, sourceName: summary.sourceName),
            tiles: summary.tiles.map(\.id),
            progress: summary.progress
        )
    }

    /// Where a card leads.
    ///
    /// A shelf the reader made opens its own detail screen, as it does from the shelves
    /// screen. A shelf a server defined opens that server's page for it, which is where an
    /// unreachable server is reported — and reporting it there is what keeps this surface
    /// silent about the network.
    ///
    /// A remembered shelf whose source this device cannot open is not on this surface at all,
    /// so the last branch is unreachable; it is written rather than trapped, because a view
    /// that cannot be built is a crash and an empty one is a dead end the reader can leave.
    @ViewBuilder
    private func destination(of summary: HomeShelfSummary) -> some View {
        switch summary.destination {
        case let .onDevice(id):
            if summary.kind == .readingList {
                ReadingListDetail(model: model, id: id)
            } else {
                CollectionDetail(model: model, id: id)
            }

        case let .onServer(shelf):
            if let page = pages[shelf.sourceID] {
                if shelf.kind == .readingList {
                    KavitaListView(server: page, listID: shelf.serverID, title: shelf.title, onOpen: onOpen)
                } else {
                    KavitaCollectionView(
                        server: page,
                        collectionID: shelf.serverID,
                        title: shelf.title,
                        onOpen: onOpen
                    )
                }
            } else {
                EmptyView()
            }
        }
    }

}

/// What a shelf's card says under its name: where it came from, and how much is in it.
///
/// One function for the home surface and the shelves screen, because it is one sentence. A
/// remembered shelf has no count this device can vouch for, so it states its source alone
/// rather than a zero — and a shelf the reader made here has no source, so it states the
/// count alone.
func shelfSubtitle(count: Int?, sourceName: String?) -> String {
    let items = count.map { count in
        String(
            format: String(localized: "shelves.count \(count)", bundle: .module, locale: .storyArc),
            count
        )
    }
    return [sourceName, items].compactMap { $0 }.joined(separator: " · ")
}

extension KavitaPage {
    /// Every Kavita server this device can actually open, by source id.
    ///
    /// Read once rather than per card: the key comes out of the secure store, and a computed
    /// property on a view would ask it again on every redraw. A source that is not a Kavita
    /// server, has no address, or has lost its key is absent — which is also what decides
    /// whether its remembered shelves are named at all, so the two are one lookup.
    static func pages(in registry: SourceRegistry, credentials: CredentialStore?) -> [UUID: KavitaPage] {
        var found: [UUID: KavitaPage] = [:]
        for source in registry.sources {
            if let page = KavitaPage(source: source, credentials: credentials) {
                found[source.id] = page
            }
        }
        return found
    }
}

File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/HomeShelvesRow.kt
package app.storyarc.feature.library

import android.graphics.Bitmap
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcSpace
import app.storyarc.core.model.Publication

/**
 * One of the home surface's two shelves of shelves.
 *
 * `collections-and-reading-lists`, *Shelves on the home surface*. Absent when its half holds
 * nothing, by the rule [shelf] already follows for a run of covers -- a heading over a gap is
 * the surface looking like it is waiting for something, which is the one thing it must never
 * do.
 *
 * **The heading leads to the shelves screen, not to a filtered library.** Every other heading
 * here leads to "the full list in the library, filtered to match the shelf", and a collection
 * is not a library filter: the exhaustive list of collections is a screen of its own, which is
 * also the only place a shelf is made, renamed or deleted. [pinnedShelves] makes the same
 * argument from the other side, where the shelf is a collection's *contents*.
 *
 * iOS's `homeShelvesSection` is the same shelf.
 */
internal fun LazyListScope.homeShelvesShelf(
    heading: Int,
    summaries: List<HomeShelfSummary>,
    cover: suspend (Publication, Int) -> Bitmap?,
    onOpenShelf: (HomeShelfSummary) -> Unit,
    onShowAll: () -> Unit,
) {
    if (summaries.isEmpty()) return
    item(key = "shelves-heading-$heading") { HomeHeading(heading, onShowAll) }
    item(key = "shelves-run-$heading") {
        HomeShelfRun(summaries = summaries, cover = cover, onOpenShelf = onOpenShelf)
    }
}

/** A run of shelves at the width four covers still read as four. */
@Composable
private fun HomeShelfRun(
    summaries: List<HomeShelfSummary>,
    cover: suspend (Publication, Int) -> Bitmap?,
    onOpenShelf: (HomeShelfSummary) -> Unit,
) {
    // The shelves screen's own floor, not a cover cell's: a shelf is a composite of four
    // covers, and four covers below about 150 dp stop being four covers. Stepped for the text
    // size by the same ladder, because the caption under a shelf cramps the same way.
    val width = shelfLatticeMinimumWidth(LocalDensity.current.fontScale)
    LazyRow(
        contentPadding = PaddingValues(horizontal = StoryArcSpace.gutter),
        horizontalArrangement = Arrangement.spacedBy(StoryArcSpace.rowCoverGap),
        modifier = Modifier.fillMaxWidth(),
    ) {
        items(summaries, key = { it.key }) { summary ->
            HomeShelfLink(
                summary = summary,
                cover = cover,
                width = width,
                modifier = Modifier
                    .width(width)
                    .clickable { onOpenShelf(summary) },
            )
        }
    }
}

/**
 * One shelf on the home surface: its artwork, then its name, then what it is.
 *
 * The same three parts, in the same order, as the shelves screen's `ShelfCard`, and drawing
 * them here rather than calling that card is one divergence with a reason: `ShelfCard` resolves
 * its own artwork through a [LibraryViewModel], and the home surface deliberately holds no view
 * model -- `HomeScreen`'s signature is its promise that the surface is assembled from local
 * data it was handed. What is *not* duplicated is everything that decides how a shelf looks:
 * [ShelfComposite] lays out the quadrants and [ShelfProgressRail] draws the rail, for both.
 */
@Composable
private fun HomeShelfLink(
    summary: HomeShelfSummary,
    cover: suspend (Publication, Int) -> Bitmap?,
    width: Dp,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current
    val caption = homeShelfCaption(summary)

    Column(
        modifier = modifier.clearAndSetSemantics {
            contentDescription = listOf(summary.name, caption).filter { it.isNotEmpty() }
                .joinToString(", ")
        },
    ) {
        Box {
            HomeShelfArtwork(tiles = summary.tiles, cover = cover, width = width)
            summary.fraction?.let { ShelfProgressRail(it) }
        }
        Column(modifier = Modifier.padding(top = StoryArcSpace.sm)) {
            Text(
                text = summary.name,
                style = MaterialTheme.typography.bodyMedium,
                color = palette.textPrimary,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
            if (caption.isNotEmpty()) {
                Text(
                    text = caption,
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.textTertiary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

/**
 * A shelf's artwork, resolved through the home surface's own cover loader.
 *
 * [ShelfComposite] decides what the frame holds; this only fetches. A shelf a server defined
 * has no local members, so it has no tiles and the composite draws the cover-shaped blank --
 * which is the case this surface meets most, and is why the blank had to stay a blank rather
 * than become a folder glyph.
 */
@Composable
private fun HomeShelfArtwork(
    tiles: List<Publication>,
    cover: suspend (Publication, Int) -> Bitmap?,
    width: Dp,
) {
    val density = LocalDensity.current
    val maxPixelSize = remember(density, width) { with(density) { width.roundToPx() } }
    val covers = remember { mutableStateMapOf<String, Bitmap>() }

    LaunchedEffect(tiles) {
        for (publication in tiles) {
            if (covers.containsKey(publication.id)) continue
            cover(publication, maxPixelSize)?.let { covers[publication.id] = it }
        }
    }

    ShelfComposite(tiles = tiles.map { it.id }, covers = covers)
}

/**
 * What a card says under its name: where the shelf came from, and how much is in it.
 *
 * The shelves screen's caption, asked the same way. A remembered shelf has no count this
 * device can vouch for, so it states its source alone rather than a zero -- and a shelf the
 * reader made here has no source, so it states the count alone.
 */
@Composable
private fun homeShelfCaption(summary: HomeShelfSummary): String {
    val items = summary.count?.let { pluralStringResource(R.plurals.shelves_count, it, it) }
    return listOfNotNull(summary.sourceName, items).joinToString(" · ")
}
